src/install-peerdeps.js

```diff
@@ -130,152 +153,176 @@ function installPeerDeps(
   cb
 ) {
   const encodedPackageName = encodePackageName(packageName);
-  getPackageData({ encodedPackageName, registry, auth, proxy })
-    // Catch before .then because the .then is so long
-    .catch(err => cb(err))
-    .then(data => {
-      const versions = Object.keys(data.versions);
-      // Get max satisfying semver version
-      let versionToInstall = maxSatisfying(versions, version);
-      // If we didn't find a version, maybe it's a tag
-      if (versionToInstall === null) {
-        const tags = Object.keys(data["dist-tags"]);
-        //  If it's not a valid tag, throw an error
-        if (tags.indexOf(version) === -1) {
-          return cb(new Error("That version or tag does not exist."));
-        }
-        // If the tag is valid, then find the version corresponding to the tag
-        versionToInstall = data["dist-tags"][version];
-      }
+  readPackageFile()
+    .then(localPackageFile =>
+      getPackageData({ encodedPackageName, registry, auth, proxy })
```

Comment:
One way we can flatten this Promise tree is something like this:

```js
let localPackageFile;
readPackageFile()
	.then(l => {
		localPackageFile = l;
		return getPackageData({ encodedPackageName, registry, auth, proxy });
	})
	.then(data => {
		// ...
		// body of function
		// ...
	})
	.catch(err => cb(err));
```

I think it'll make the function a bit easier to read/reason about!

Reply:
Personally I'm not keen on setting variables outside the promise callbacks.

Normally these days I use async/await instead of raw promises. You should consider using async/await if you want it easier to read and reason about ;)

Reply:
I agree -- async/await is a beautiful abstraction over callbacks and Promises!

Unfortunately, there's a project at Airbnb which depends on this project maintaining compatibility with Node.js 0.10 and 0.12, and I don't want to use async/await yet as I'm not completely sure if `regeneratorRuntime` has dropped support for earlier Node versions -- so that's why we're staying with Promises for now.